Source/WebKit/WebProcess/WebCoreSupport/WebNotificationClient.cpp

```diff
@@ -102,41 +111,46 @@ void WebNotificationClient::requestPermission(ScriptExecutionContext& context, P
     if (!securityOrigin)
         return permissionHandler(NotificationClient::Permission::Denied);
 
-    // Add origin to list of origins that have requested permission to use the Notifications API.
-    m_notificationPermissionRequesters.add(securityOrigin->data());
-
     m_page->notificationPermissionRequestManager()->startRequest(securityOrigin->data(), WTFMove(permissionHandler));
 }
 
 NotificationClient::Permission WebNotificationClient::checkPermission(ScriptExecutionContext* context)
 {
     if (!context
         || (!context->isDocument() && !context->isServiceWorkerGlobalScope())
+        || !context->securityOrigin()
         || WebProcess::singleton().sessionID().isEphemeral())
         return NotificationClient::Permission::Denied;
 
-    auto* origin = context->securityOrigin();
-    if (!origin)
+    auto source = Permissions::sourceFromContext(*context);
+    if (!source)
         return NotificationClient::Permission::Denied;
 
-    NotificationClient::Permission resultPermission;
-    callOnMainRunLoopAndWait([&resultPermission, origin = origin->data().toString().isolatedCopy()] {
-        resultPermission = WebProcess::singleton().supplement<WebNotificationManager>()->policyForOrigin(origin);
+    Page* page = nullptr;
+    if (source == WebCore::PermissionQuerySource::Window || source == WebCore::PermissionQuerySource::DedicatedWorker)
+        page = m_page->corePage();
+
+    auto resultPermission = NotificationClient::Permission::Default;
+    callOnMainRunLoopAndWait([&resultPermission, topOrigin = crossThreadCopy(context->topOrigin().data()), securityOrigin = crossThreadCopy(context->securityOrigin()->data()), page, source = *source] {
+        PermissionController::shared().query(ClientOrigin { topOrigin, securityOrigin }, PermissionDescriptor { WebCore::PermissionName::Notifications }, page, source, WebCore::Asynchronous::No, [&resultPermission] (auto permissionState) mutable {
+            if (permissionState) {
+                switch (*permissionState) {
+                case WebCore::PermissionState::Granted:
+                    resultPermission = NotificationClient::Permission::Granted;
+                    break;
+                case WebCore::PermissionState::Denied:
+                    resultPermission = NotificationClient::Permission::Denied;
+                    break;
+                default: 
```

Comment:
The only default case is PermissionState::Prompt so you may just use PermissionState::Prompt for clarity and set resultPermission accordingly.